dbs-upcall: introduce upcall

studychao opened the pull request:
dbs-upcall is a direct communication tool between VMM and guest
developed upon vsock. The server side of the upcall is a driver in guest
kernel (kernel patches are needed for this feature) and it'll start to
serve the requests after the kernel starts. And the client side is in
VMM , it is a thread that communicates with vsock through uds.

We have accomplished device hotplug / hot-unplug directly through upcall
in order to avoid virtualization of ACPI to minimize virtual machines'
overhead. And there could be many other usage through this direct
communication channel.

Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>

studychao committed:
dbs-upcall: introduce upcall

dbs-upcall is a direct communication tool between VMM and guest
developed upon vsock. The server side of the upcall is a driver in guest
kernel (kernel patches are needed for this feature) and it'll start to
serve the requests after the kernel starts. And the client side is in
VMM, it is a thread that communicates with vsock through uds.

We have accomplished device hotplug / hot-unplug directly through upcall
in order to avoid virtualization of ACPI to minimize virtual machines'
overhead. And there could be many other usage through this direct
communication channel.

Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>

ZizhengBian commented on crates/dbs-upcall/Cargo.toml:
```diff
@@ -0,0 +1,20 @@
+
+[dependencies]
+anyhow = "1"
+slog = "2"
```
Use standard logger is better?

studychao replied:
changed. Please have a look.

ZizhengBian commented on crates/dbs-upcall/kernel/0001-dragonball-introduce-dragonball-subarch.patch:
```diff
@@ -0,0 +1,854 @@
+Date: Tue, 10 May 2022 17:09:40 +0800
+Subject: [PATCH 1/2] dragonball: introduce dragonball subarch
+
+Dragonball subarch is to let guest know it is in dragonball guest so
```
subarch havs wake relationship with upcall, should we move our kernel patch to a more common folder?

studychao replied:
Changed.

ZizhengBian commented on crates/dbs-upcall/kernel/0001-dragonball-introduce-dragonball-subarch.patch:
```diff
@@ -0,0 +1,854 @@
++	bool stable;
++	bool state;
++	u64  features;
++} dragonball_priv __read_mostly;
```
Since we already have dragonball platfrom device, so can refactor the code to make upcall vsock server to a common platform feature.

ZizhengBian replied:
And I think dragonball_priv can be replaced by struct dragonball_platform

ZizhengBian commented on crates/dbs-upcall/kernel/0001-dragonball-introduce-dragonball-subarch.patch:
```diff
@@ -0,0 +1,854 @@
++} features[] = {
++	DRAGONBALL_FEATURES
++};
++#undef DB_ADD_FEATURE
```
Do you think here is over design, I suggest use a more simple way to achieve this

ZizhengBian commented on crates/dbs-upcall/kernel/0001-dragonball-introduce-dragonball-subarch.patch:
```diff
@@ -0,0 +1,854 @@
++#define DB_ADD_FEATURE(_feat, _name)					\
++	[DB_FEAT_##_feat] = {						\
++		.name = "dragonball.no"#_name,				\
++		.val = (1 << DB_FEAT_##_feat),				\
```
If we have dragonball platform device, do we need to use cmdline to disable featrues?

ZizhengBian commented on crates/dbs-upcall/kernel/0002-upcall-add-vsock-server-and-upcall-support.patch:
```diff
@@ -0,0 +1,1145 @@
+--- a/drivers/misc/dragonball/Kconfig
++++ b/drivers/misc/dragonball/Kconfig
+@@ -15,3 +15,7 @@ menuconfig DRAGONBALL_DRIVERS
+ 	  add any kernel code.
```
do you miss some code?

ZizhengBian commented on crates/dbs-upcall/kernel/0002-upcall-add-vsock-server-and-upcall-support.patch:
```diff
@@ -0,0 +1,1145 @@
++
++/* Type of request and reply messages. */
++enum devmgr_msg_type {
++	CONNECT         = 0x00000000,
```
tab indent is incorrect

ZizhengBian commented on crates/dbs-upcall/kernel/0002-upcall-add-vsock-server-and-upcall-support.patch:
```diff
@@ -0,0 +1,1145 @@
++
++config DRAGONBALL_HOTPLUG_CPU
++	bool "CPU hotplug/hotunplug support"
++	depends on DRAGONBALL_DEVICE_MANAGER
```
add depends `x86_64`?

ZizhengBian commented on crates/dbs-upcall/kernel/0002-upcall-add-vsock-server-and-upcall-support.patch:
```diff
@@ -0,0 +1,1145 @@
++	pr_info("cpu event notification: apic ids index %d", apic_ids_index);
++	rep->msg_load.cpu_dev_info.apic_index = apic_ids_index;
++	rep->ret = ret;
++	_fill_msg_header(&rep->msg_header,
```
Can we simplify this function, e.g. just return error code, and fill msg header in a more common function.

ZizhengBian commented on crates/dbs-upcall/kernel/0002-upcall-add-vsock-server-and-upcall-support.patch:
```diff
@@ -0,0 +1,1145 @@
++	}
++
++	while (!kthread_should_stop()) {
++		len = db_vsock_recvmsg(res->sock, (char *)&res->req,
```
should we ensure that res->req meet the length of DEV_MGR_MSG_SIZE？

ZizhengBian commented on crates/dbs-upcall/kernel/0002-upcall-add-vsock-server-and-upcall-support.patch:
```diff
@@ -0,0 +1,1145 @@
++
++static int __init db_device_manager_init(void)
++{
++	if (has_dragonball_feature(DB_FEAT_DEVMGR))
```
I suggest using platform device's capability probe instead of its own init function.

ZizhengBian commented on crates/dbs-upcall/kernel/0002-upcall-add-vsock-server-and-upcall-support.patch:
```diff
@@ -0,0 +1,1145 @@
++	return 0;
++}
++
++late_initcall(db_vsock_srv_init);
```
I suggest using platform device's capability probe instead of its own init function.

ZizhengBian commented on crates/dbs-upcall/kernel/0001-upcall-add-vsock-server-and-upcall-support.patch:
```diff
@@ -0,0 +1,1178 @@
++#include <linux/net.h>
++
++/* Vsock port to listen for incoming connections. */
++#define DB_SERVER_PORT			0xDB
```
tab indent?

ZizhengBian commented on crates/dbs-upcall/README.md:
```diff
@@ -0,0 +1,70 @@
+# dbs-upcall
+
```
Add DB_SERVER_PORT in doc?

studychao replied:
Added

ZizhengBian commented on crates/dbs-upcall/kernel/0001-upcall-add-vsock-server-and-upcall-support.patch:
```diff
@@ -0,0 +1,1178 @@
+---
+ drivers/misc/Kconfig                          |   1 +
+ drivers/misc/Makefile                         |   1 +
+ drivers/misc/dragonball/Kconfig               |  21 +
```
Should we split dragonball and upcall's code into two patches?

studychao replied:
Changed.

ZizhengBian commented on crates/dbs-upcall/kernel/0001-upcall-add-vsock-server-and-upcall-support.patch:
```diff
@@ -0,0 +1,1178 @@
++
++/* These definitions are synchronized with dragonball */
++#define DEV_MGR_MSG_SIZE        0x400
++#define DEVMGR_CMD_BYTE			'd'
```
tab indent

studychao replied:
Changed.

ZizhengBian commented on crates/dbs-upcall/kernel/0001-upcall-add-vsock-server-and-upcall-support.patch:
```diff
@@ -0,0 +1,1178 @@
++/* Type of request and reply messages. */
++enum devmgr_msg_type {
++	CONNECT         = 0x00000000,
++	ADD_CPU			= 0x00000001,
```
tab indent

studychao replied:
Changed.

studychao committed:
kernel: introduce dragonball driver patch

Dragonball driver patch in kernel is a prerequisite for our more other
features.

Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>

studychao committed:
kernel: introduce upcall kernel patch

upcall kernel patch is introduced to enable upcall in the dragonball
driver.

Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>

ZizhengBian commented on kernel/0001-dragonball-introduce-dragonball-driver.patch:
```diff
@@ -0,0 +1,75 @@
++#
++# Makefile for Dragonball misc drivers
++#
+\ No newline at end of file
```
need a blank line here, pls use checkpatch.pl check your patch

studychao replied:
Used.

ZizhengBian commented on crates/dbs-upcall/README.md:
```diff
@@ -8,7 +8,7 @@ We have accomplished device hotplug / hot-unplug directly through upcall in orde
 ### Server Design
 
-The server side of upcall is a driver in guest kernel.
+The server side of upcall is a driver in guest kernel and the vsock port is 0xDB.
```
Add port and kernel's url to rust's doc, too?

studychao replied:
Already Added.

studychao committed:
README: add vsock port for upcall document

add vsock port in upcall document to help users.

Signed-off-by: Chao Wu <chaowu@linux.alibaba.com>

ZizhengBian reviewed (approved):
LGTM
